docs: add guidance about names of workload-less charms (#2496)
canonical/operator: docs/howto/initialise-your-project.md

Reviewer: suggestion: I would also mention the distinction between normal operators, whose repository suffixes are not included in charm names, and integrators and configurators, whose charm suffixes are often (but not always, hence the necessary disambiguation) also included in charm names, instead - for example: [`kubeflow-dashboard-operator`](https://github.com/canonical/kubeflow-dashboard-operator/blob/main/metadata.yaml#L3) vs [`data-kubeflow-integrator`](https://github.com/canonical/data-kubeflow-integrator/blob/main/metadata.yaml#L4)

```text
https://charmhub.io/mega-calendar-k8s
```

> There also exist charms that do not operate a workload, such as integrators and configurators charms. These two categories serve two different purposes:

Reviewer 2: Nit:

```suggestion
Some charms do not operate a workload, such as integrator and configurator charms. These categories serve different purposes:
```

> * An integrator charm provides the possibility to integrate a service that is not managed via Juju into the Juju model. This can both apply to server side integrations (e.g. `s3-integrator`, that integrates an externally managed s3 object storage) or to client side integration (e.g. `data-integrator`, representing the integration of external client applications that needs a database).

Reviewer 2: ```suggestion
* An integrator charm alows integration with a service that is not managed via Juju into the Juju model. This can both apply to server side integrations (such as `s3-integrator`, which integrates an externally managed S3 object storage) or to client side integration (such as `data-integrator`, representing the integration of external client applications that needs a database).
```

Reviewer 3: We try to avoid "via" - plus a couple of typo fixes:

```suggestion
* An integrator charm allows integration with a service that is not managed through Juju. This can both apply to server side integrations (such as `s3-integrator`, which integrates an externally managed S3 object storage) or to client side integrations (such as `data-integrator`, representing the integration of external client applications that needs a database).
* 
```

If you feel that dropping "into the Juju model" is going too far, fair enough (again!)

> * A configurator charm provides better scalability and centralized logic to further configure for a particular charm or relation that is already in Juju. Examples for this could be `cos-configuration` when it applies to a single charm (e.g. providing more fine-grained configuration the prometheus scraping) or for relation/integration (e.g. `ingress-configurator` as to provide further configuration of ingresses requests)

Reviewer 2: I believe we have not yet moved these docs to en-US. Some other small tweaks:

```suggestion
* A configurator charm provides additional scalability and centralised logic to configure a particular charm or relation that is already in Juju. Examples for this include `cos-configuration` when it applies to a single charm (such as providing more fine-grained configuration of the prometheus scraping) or for a relation (such as `ingress-configurator` to provide additional configuration of ingress requests).
```

Reviewer 3: I agree. Building on the suggestion:

```suggestion
* A configurator charm provides logic to configure a particular charm or relation that is already in Juju. Examples include `cos-configuration` when it applies to a single charm (such as providing more fine-grained configuration of the Prometheus scraping) or for a relation (such as `ingress-configurator` to provide additional configuration of ingress requests).
```

If you feel that dropping "additional scalability and centralised" is going too far, OK, fair enough 🙂

> Since workload-less charms can work both on machines and on Kubernetes, when naming integrator charms and configurator charms, avoid using the `k8s` suffix, unless the charm is only relevant for Kubernetes, e.g. managing K8s resources within the charm logic using `lightkube`. When naming the charm, use the `integrator` and `configurator` suffix to signal the category of the charm, e.g. `foo-integrator` or `bar-configurator`.

Reviewer 2: We avoid "e.g.". Also, I think the last sentence is a separate point.

```suggestion
Since workload-less charms can work both on machines and on Kubernetes, avoid using the `k8s` suffix when naming integrator charms and configurator charm , unless the charm is only relevant for Kubernetes (for example, managing K8s resources within the charm logic using `lightkube`).

When naming the charm, use the `integrator` and `configurator` suffix to signal the category of the charm. For example, `foo-integrator` or `bar-configurator`.
```

> (create-a-repository)=

Reviewer 2: I'm a bit concerned about changing the anchor name. I can't find links to it in our docs (other than charmhub-listing-review and the best practice list, which will auto-fix), but we are potentially breaking someone's links (if it's just a fragment link in a full URL it's probably ok, people end up on the page and can search, but intersphinx links will break).

I think we could adjust the content as suggested but keep it under one heading.

Reviewer 3: This comes from my idea to split "Create a repository and initialise it" into two headings, which I feel works better. I searched for other projects using the intersphinx target and couldn't find any, so I felt it would be safe to change.

(Other than breaking the best practice listing, which is what is causing the docs build to fail. @deusebio, don't worry about that - one of us will be able to fix it. Or see the docstring of `.github/update-best-practice-table.py` if you want to fix it yourself.)

I prefer having two headings, but I don't mind if you prefer to switch back to one. "Initialise the repository" will get shorter in the future when we remove tip about using Charmcraft from GitHub.

Reviewer 2: Ok, I also couldn't find any definite breakage, so fine with me.

Author: I hope I have addressed this in the commit [2c07f0b](https://github.com/canonical/operator/pull/2496/commits/2c07f0b03370eb9a3c05ac4c56df53082a02a3f1)

## Create a repository

Create a repository with your source control of choice.

```{admonition} Best practice
:class: hint

> If your charm operates a workload, name the repository `<charm name>-operator`. If your charm doesn't operate a workload (as in the case of integrator charms and configurator charms), the `-operator` suffix is not needed and the repository can be named with the same name of the charm, e.g. `foo-integrator` and `bar-configurator`. Use the plural when `s` the repository contains multiple charms or artefacts (such as rocks).

Reviewer 2: I'd still like to make the multiple artefacts case more vague. If it's multiple charms, probably put an 's' on the end, but if it's rocks and other things, probably the workload name, if it's a monorepo then something else - it gets too complicated to include here, I feel.

```suggestion
If your charm operates a workload, name the repository `<charm name>-operator`. If your charm doesn't operate a workload (as in the case of integrator charms and configurator charms), the `-operator` suffix is not needed and the repository can be named with the same name of the charm, for example `foo-integrator` and `bar-configurator`.

Repositories that contain multiple charms or one or more charms and other artefacts (like Rocks) will need to use other naming patterns.
```

Reviewer 3: I agree with Tony's comment. Adding my own suggestions on top:

```suggestion
If your charm operates a workload, name the repository `<charm name>-operator`. For advice about the charm name, see [](#decide-your-charms-name). If your charm doesn't operate a workload (as in the case of integrator charms and configurator charms), the `-operator` suffix isn't needed. For example, `foo-integrator` and `bar-configurator`.

Repositories that contain multiple charms or one or more charms and other artefacts (like Rocks) will need to use other naming patterns.
```

```

Examples:

- [kafka-operator](https://github.com/canonical/kafka-operator) - Contains a single charm that operates a workload on machine.
- [kafka-k8s-operator](https://github.com/canonical/kafka-k8s-operator) - Contains a single charm that operates a workload on K8s.
- [katib-operators](https://github.com/canonical/katib-operators) - Contains multiple charms.

> - [data-integrator](https://github.com/canonical/data-integrator) - Contains a charm that integrates an externally managed service (e.g. client application).

Reviewer 2: ```suggestion
- [data-integrator](https://github.com/canonical/data-integrator) - Contains a charm that integrates an externally managed service (such as a client application).
```

> - [s3-integrator](https://github.com/canonical/object-storage-integrators) - Contains multiple charm that integrates externally managed service (e.g. different kind of object storage backends).

Reviewer 2: ```suggestion
- [s3-integrator](https://github.com/canonical/object-storage-integrators) - Contains multiple charms that integrate an externally managed service (such as different kinds of object storage backends).
```

> - [request-authentication-configurator](https://github.com/canonical/request-authentication-configurator) - Contains a charm that configure Gateway to perform request authentication

Reviewer 2: ```suggestion
- [request-authentication-configurator](https://github.com/canonical/request-authentication-configurator) - Contains a charm that configures Gateway to perform request authentication.
```

(initialise-the-repository)=

> ## Initialise the repository

Reviewer 2: I think we do want to keep the link to picking a name.

```suggestion
## Initialise the repository

To appropriately name the charm, see [](#decide-your-charms-name).
```

Or maybe it belongs up higher than the repo name, as something done first?

Reviewer 3: This was another of my ideas. My thinking was, the generated best practice bullet ends a bit weirdly:

> [...] see [Decide your charm's name](https://documentation.ubuntu.com/ops/latest/howto/initialise/-your-project/#decide-your-charms-name). See [Create a repository and initialise it](https://documentation.ubuntu.com/ops/latest/howto/initialise-your-project/#create-a-repository-and-initialise-it).

How about we keep the link in the best practice note, but move it closer to the beginning. I've just added a suggestion about that.

So I suggest we don't do what Tony is suggesting in this comment. Outside the context of the best practice bullet, the "Decide your charm’s name" heading should be discoverable enough from the page layout and TOC.

```text
https://charmhub.io/mega-calendar-k8s
```

Some charms do not operate a workload, such as integrator and configurator charms. These categories serve different purposes:

> * An integrator charm allows integration with a service that is not managed through Juju. This can both apply to server side integrations (such as `s3-integrator`, which integrates an externally managed S3 object storage) or to client side integrations (such as `data-integrator`, representing the integration of external client applications that needs a database).

Reviewer 3: ```suggestion
* An integrator charm allows integration with a service that is not managed through Juju. This can both apply to server side integrations (such as `s3-integrator`, which integrates an externally managed S3 object storage) or to client side integrations (such as `data-integrator`, representing the integration of external client application that needs a database).
```

> * 

Reviewer 3: ```suggestion
```

* A configurator charm provides logic to configure a particular charm or relation that is already in Juju. Examples include `cos-configuration` when it applies to a single charm (such as providing more fine-grained configuration of the Prometheus scraping) or for a relation (such as `ingress-configurator` to provide additional configuration of ingress requests).

> Since workload-less charms can work both on machines and on Kubernetes, avoid using the `k8s` suffix when naming integrator charms and configurator charm , unless the charm is only relevant for Kubernetes (for example, managing K8s resources within the charm logic using `lightkube`).

Reviewer 3: ```suggestion
Since workload-less charms can work both on machines and on Kubernetes, avoid using the `k8s` suffix when naming integrator charms and configurator charms, unless the charm is only relevant for Kubernetes (for example, managing K8s resources within the charm logic using `lightkube`).
```

> When naming the charm, use the `integrator` and `configurator` suffix to signal the category of the charm. For example, `foo-integrator` or `bar-configurator`.

Reviewer 3: ```suggestion
When naming the charm, use `-integrator` and `-configurator` to signal the category of the charm. For example, `foo-integrator` or `bar-configurator`.
```

(create-a-repository)=
## Create a repository

Create a repository with your source control of choice.

```{admonition} Best practice
:class: hint

If your charm operates a workload, name the repository `<charm name>-operator`. For advice about the charm name, see [](#decide-your-charms-name). If your charm doesn't operate a workload (as in the case of integrator charms and configurator charms), the `-operator` suffix isn't needed. For example, `foo-integrator` and `bar-configurator`.

> Repositories that contain multiple charms or one or more charms and other artefacts (like Rocks) will need to use other naming patterns.

Reviewer 3: ```suggestion
Repositories that contain multiple charms or one or more charms and other artefacts (like rocks) will need to use other naming patterns.
```

```

Examples:

- [kafka-operator](https://github.com/canonical/kafka-operator) - Contains a single charm that operates a workload on machine.

> - [kafka-k8s-operator](https://github.com/canonical/kafka-k8s-operator) - Contains a single charm that operates a workload on K8s.

Reviewer 3: Redoing this suggestion as my previous attempt doesn't look right

```suggestion
- [kafka-operator](https://github.com/canonical/kafka-operator) - Contains a single charm that operates a machine workload.
- [kafka-k8s-operator](https://github.com/canonical/kafka-k8s-operator) - Contains a single charm that operates a K8s workload.
```

```{admonition} Best practice
:class: hint

> Name the repository using the pattern ``<charm name>-operator`` for a single charm managing a workload. Workloadless charms managing integration with another charm should use the suffix ``integrator`` instead. Charms managing configuration for another charm should use ``configurator``. Monorepos containing multiple related charms should instead use a plural suffix, like ``operators``. For the charm name, see [](#decide-your-charms-name).

Reviewer: As confirmed by @james-garner-canonical, we could avoid this ambiguity by:

```suggestion
Name the repository using the pattern ``<charm name>-operator`` for a single charm managing a workload. Workloadless charms managing integration with another charm should use the suffix ``integrator`` instead. Workloadless charms managing configuration for another charm should use ``configurator``. Monorepos containing multiple related charms should instead use a plural suffix, like ``operators``. For the charm name, see [](#decide-your-charms-name).
```

Reviewer 2: I suggest we break this out and also tidy up the plural case at the same time.

```suggestion
The name of a repository containing a single charm should start with the name of the charm, and then add an "-operator" (workload charm), "-integrator" (workload-less charm managing integrations with another charm), or "-configurator" (workload-less charm managing configuration of another charm) suffix, unless the charm name already contains the suffix.

Repositories that contain multiple charms or one or more charms and other artefacts (like Rocks) will need to use other naming patterns.

For the charm name, see [](#decide-your-charms-name).
```

And then I think we should expand the examples section below to include at least one of each of these cases, for real repositories that we think use the right pattern.

Name the repository using the pattern ``<charm name>-operator`` for a single
charm, or ``<base charm name>-operators`` when the repository will hold
multiple related charms. For the charm name, see [](#decide-your-charms-name).

> The suffix `-operator` applies only to charms that manage workloads and operate a particular software. Do not use "-operator" suffix for workless charm (e.g. integrator charms) that do not manage any workload for avoiding suffix stacking and favoring a leaner naming, such as `data-integrator` instead of `data-integrator-operator`. 

Reviewer 4: Thanks for opening a PR, avoiding convoluted suffix stacking sounds great to us. Having conventional repository names is handy for discovery, so what do you think about enumerating the suffixes here? Besides 'operator' and 'integrator', I guess there's also 'configurator'. Are there more we should document?

I think this can replace the guidance we documented originally, perhaps like this:

```suggestion
Name the repository using the pattern ``<charm name>-operator`` for a single charm managing a workload. Workloadless charms managing integration with another charm should use the suffix ``integrator`` instead. Charms managing configuration for another charm should use ``configurator``. Monorepos containing multiple related charms should instead use a plural suffix, like ``operators``. For the charm name, see [](#decide-your-charms-name).
```

Author: reads good!
